openshift/machine-config-operator
# OCPNODE-1632: Support Cluster ImagePolicy CRD

**- What I did**
Follow the design from the [enhancement](https://github.com/openshift/enhancements/blob/master/enhancements/api-review/add-ClusterImagePolicy-and-ImagePolicy-for-Signature-Verification.md). Add the implementation to ContainerruntimConfig controller for policy.json configuration.
**- How to verify it**
- verify on cluster with techpriview enabled

```yaml
apiVersion: config.openshift.io/v1alpha1
kind: ClusterImagePolicy 
metadata:
  name: p0
spec:
  scopes:
    - registry.build03.ci.openshift.org
    - registry.ci.openshift.org
    - example.com/global
  policy:
    rootOfTrust:
      policyType: PublicKey
      publicKey:
        keyData: Zm9vIGJhcg==
    signedIdentity:
      matchPolicy: MatchRepoDigestOrExact
```
```bash
oc create -f /home/qiwan/Documents/test-crds/hands-on-sigstore/clusterimgpolicy.yaml
#check policy.json and /etc/containers/registries.d/sigstore-registries.yaml 
$ oc debug node/ip-10-0-53-139.ec2.internal
Starting pod/ip-10-0-53-139ec2internal-debug-5rftt ...
To use host binaries, run `chroot /host`
Pod IP: 10.0.53.139
If you don't see a command prompt, try pressing enter.
sh-4.4# chroot /host
sh-5.1# cat /etc/containers/policy.json 
{"default":[{"type":"insecureAcceptAnything"}],"transports":{"atomic":{"example.com/global":[{"type":"sigstoreSigned","keyData":"Zm9vIGJhcg==","signedIdentity":{"type":"matchRepoDigestOrExact"}}],"registry.ci.openshift.org":[{"type":"sigstoreSigned","keyData":"Zm9vIGJhcg==","signedIdentity":{"type":"matchRepoDigestOrExact"}}]},"docker":{"example.com/global":[{"type":"sigstoreSigned","keyData":"Zm9vIGJhcg==","signedIdentity":{"type":"matchRepoDigestOrExact"}}],"registry.ci.openshift.org":[{"type":"sigstoreSigned","keyData":"Zm9vIGJhcg==","signedIdentity":{"type":"matchRepoDigestOrExact"}}]},"docker-daemon":{"":[{"type":"insecureAcceptAnything"}]}}}sh-5.1# 
sh-5.1# cat /etc/containers/registries.d/sigstore-registries.yaml 
docker:
  example.com/global:
    use-sigstore-attachments: true
  registry.ci.openshift.org:
    use-sigstore-attachments: true

$ oc describe clusterimagepolicy.config.openshift.io/p0
Name:         p0
Namespace:    
Labels:       <none>
Annotations:  <none>
API Version:  config.openshift.io/v1alpha1
Kind:         ClusterImagePolicy
Metadata:
  Creation Timestamp:  2024-02-29T05:05:19Z
  Generation:          1
  Resource Version:    63708
  UID:                 b18c7cf6-995d-47e7-80a0-4780ed7c52a5
Spec:
  Policy:
    Root Of Trust:
      Policy Type:  PublicKey
      Public Key:
        Key Data:  Zm9vIGJhcg==
    Signed Identity:
      Match Policy:  MatchRepoDigestOrExact
  Scopes:
    registry.build03.ci.openshift.org
    registry.ci.openshift.org
    example.com/global
Status:
  Conditions:
    Last Transition Time:  2024-02-29T05:05:19Z
    Message:               has conflict scope(s) ["registry.build03.ci.openshift.org"] of Openshift payload repository registry.build03.ci.openshift.org/ci-ln-ltipm3t/release, skip the scope(s)
    Observed Generation:   1
    Reason:                ConflictScopes
    Status:                True
    Type:                  Pending
Events:                    <none>

```
`registry.build03.ci.openshift.org` did not show in the `policy.json` because it is the release payload image repo.

- verify on cluster without techpreview. verify the patch does not fail the upgrade.
```
$ oc adm upgrade --force=true --allow-explicit-upgrade=true --to-image registry.ci.openshift.org/ocp/release:4.16.0-0.nightly-2024-02-29-062601

```
- verified the ClusterImagePolicy CR from the manifests directory got applied by bootstrap.
**- Description for the changelog**
